pkg/provider-local/webhook/controlplane/ensurer.go

```diff
@@ -92,6 +92,11 @@ func (e *ensurer) EnsureAdditionalProvisionFiles(_ context.Context, _ extensions
 			UpstreamHost: "europe-docker.pkg.dev",
 			MirrorHost:   "http://garden.local.gardener.cloud:5008",
 		},
+		// Enable containerd to reach registry at garden.local.gardener.cloud:5001 via HTTP.
+		{
+			UpstreamHost: "garden.local.gardener.cloud:5001",
+			MirrorHost:   "http://garden.local.gardener.cloud:5001",
+		},
```

Comment:
Could you add this to `EnsureCRIConfig` instead? The registries here are only needed for node bootstrapping, namely GNA and Hyperkube.

Please see comment above:
```
// Further registries are supposed to be added via the `EnsureCRIConfig` function (reconcile OSC).
```

Reply:
The first solution I tried was to add it to `EnsureCRIConfig` but there is an issue with it.

Looking at https://github.com/gardener/gardener/blob/d15e832232aee2a6554ed824acdc17f22947307e/pkg/provider-local/webhook/controlplane/ensurer.go#L115-L141 the `Upstream` field corresponds to the directory that is being created on /etc/containerd/certs.d. So for this use case the value should be `garden.local.gardener.cloud:5001`.

But when the OperatingSystemConfig resource goes trough validation here: https://github.com/gardener/gardener/blob/d15e832232aee2a6554ed824acdc17f22947307e/pkg/apis/extensions/validation/operatingsystemconfig.go#L135-L137
It is checked that `Upstream` matches the definition of an FQDN, but an FQDN _should_not_ contain a port and thus the validation fails. Here's the error:
```
- description: "task \"Deploying operating system specific configuration for shoot
      workers\" failed: retry failed with context deadline exceeded, last error: 1
      error occurred:\n\t* admission webhook \"validation.extensions.operatingsystemconfigs.resources.gardener.cloud\"
      denied the request: OperatingSystemConfig.extensions.gardener.cloud \"gardener-node-agent-local-ee46034b8269353b-original\"
      is invalid: spec.criConfig.containerd.registries[5].upstream: Invalid value:
      \"garden.local.gardener.cloud:5001\": a lowercase RFC 1123 subdomain must consist
      of lower case alphanumeric characters, '-' or '.', and must start and end with
      an alphanumeric character (e.g. 'example.com', regex used for validation is
      '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')\n\n"
```

My first idea was the change that I've currently made since RegistryMirrors do not go trough the same validation and there's no issue when you add the conf to the list of mirrors.
But I do agree that that's not ideal.

The best solution to this would be to "fix" the validation.
As far as I can tell when looking at the containerd docs: https://github.com/containerd/containerd/blob/main/docs/hosts.md#registry-host-namespace
A valid value for the directory name can include a port since it's expected that it can be a registry host namespace.
From the containerd docs:
`registry host namespace portion is [registry_host_name|IP address][:port]`

I have also tested that indeed this is the case. If there is a conf dir for `garden.local.gardener.cloud`, then accessing it on port 5001 gives us the same HTTPS communication issue, but if the dir is `garden.local.gardener.cloud:5001` the issue is resolved.

I'd prefer if I directly change the validation for the OperatingSystemConfig resource to match the containerd docs, but I'd like a second opinion before I make changes.

Reply:
> I'd prefer if I directly change the validation for the OperatingSystemConfig resource to match the containerd docs, but I'd like a second opinion before I make changes.

Oh yes absolutely, please 👍